lib/libxdp/xsk.c

```diff
@@ -0,0 +1,947 @@
+					  xsk_convert_xdp_flags(xsk->config.xdp_flags), 0);
+		if (err)
+			goto err_prog_load;
+		ctx->prog_fd = xdp_program__fd(ctx->xdp_prog);
+	} else {
+		ctx->prog_fd = prog_fd;
+		if (ctx->prog_fd < 0)
+			return -errno;
+		err = xsk_lookup_bpf_map(xsk);
+		if (err) {
+			close(ctx->prog_fd);
+			return err;
+		}
+	}
+
+	if (xsk->rx) {
+		err = xsk_add_map_entry(xsk);
+		if (err) {
+			if (prog_fd < 0) {
+				goto err_detach;
+			} else {
+				close(ctx->prog_fd);
+				return err;
+			}
+		}
+	}
+	if (xsks_map_fd)
+		*xsks_map_fd = ctx->xsks_map_fd;
+
+	return 0;
+
+err_detach:
+	xdp_program__detach(ctx->xdp_prog, ctx->ifindex,
+			    xsk_convert_xdp_flags(xsk->config.xdp_flags), 0);
+err_prog_load:
+	xdp_program__close(ctx->xdp_prog);
+	return err;
+}
+
+static struct xsk_ctx *xsk_get_ctx(struct xsk_umem *umem, int ifindex,
+				   __u32 queue_id)
+{
+	struct xsk_ctx *ctx;
+
+	if (list_empty(&umem->ctx_list))
+		return NULL;
+
+	list_for_each_entry(ctx, &umem->ctx_list, list) {
+		if (ctx->ifindex == ifindex && ctx->queue_id == queue_id) {
+			ctx->refcount++;
+			return ctx;
+		}
+	}
+
+	return NULL;
+}
+
+static void xsk_put_ctx(struct xsk_ctx *ctx)
+{
+	struct xsk_umem *umem = ctx->umem;
+	struct xdp_mmap_offsets off;
+	int err;
+
+	if (--ctx->refcount == 0) {
+		err = xsk_get_mmap_offsets(umem->fd, &off);
+		if (!err) {
+			munmap(ctx->fill->ring - off.fr.desc,
+			       off.fr.desc + umem->config.fill_size *
+			       sizeof(__u64));
+			munmap(ctx->comp->ring - off.cr.desc,
+			       off.cr.desc + umem->config.comp_size *
+			       sizeof(__u64));
+		}
+
+		list_del(&ctx->list);
+		free(ctx);
+	}
+}
+
+static struct xsk_ctx *xsk_create_ctx(struct xsk_socket *xsk,
+				      struct xsk_umem *umem, int ifindex,
+				      const char *ifname, __u32 queue_id,
+				      struct xsk_ring_prod *fill,
+				      struct xsk_ring_cons *comp)
+{
+	struct xsk_ctx *ctx;
+	int err;
+
+	ctx = calloc(1, sizeof(*ctx));
+	if (!ctx)
+		return NULL;
+
+	if (!umem->fill_save) {
+		err = xsk_create_umem_rings(umem, xsk->fd, fill, comp);
+		if (err) {
+			free(ctx);
+			return NULL;
+		}
+	} else if (umem->fill_save != fill || umem->comp_save != comp) {
+		/* Copy over rings to new structs. */
+		memcpy(fill, umem->fill_save, sizeof(*fill));
+		memcpy(comp, umem->comp_save, sizeof(*comp));
+	}
+
+	ctx->ifindex = ifindex;
+	ctx->refcount = 1;
+	ctx->umem = umem;
+	ctx->queue_id = queue_id;
+	memcpy(ctx->ifname, ifname, IFNAMSIZ - 1);
+	ctx->ifname[IFNAMSIZ - 1] = '\0';
+
+	umem->fill_save = NULL;
+	umem->comp_save = NULL;
+	ctx->fill = fill;
+	ctx->comp = comp;
+	list_add(&ctx->list, &umem->ctx_list);
+	return ctx;
+}
+
+static void xsk_destroy_xsk_struct(struct xsk_socket *xsk)
+{
+	free(xsk->ctx);
+	free(xsk);
+}
+
+int xsk_socket__update_xskmap(struct xsk_socket *xsk, int fd)
+{
+	xsk->ctx->xsks_map_fd = fd;
+	return xsk_add_map_entry(xsk);
+}
+
+int xsk_setup_xdp_prog(int ifindex, int *xsks_map_fd)
```

Comment:
How is the application supposed to tear down the XDP prog again? Or rather, how is an application supposed to do this without racing against another program trying to install itself? Since we abstract away the creation and attachment of the XDP program, I think we need to handle the race condition as well...

We do already have a locking mechanism (in xdp_lock_acquire/release()) we could use, but a bit of care is needed to prevent deadlocking with outselves (xdp_program__attach() will grab the lock). So may need a bit more knowledge of the internals of the rest of the library... WDYT?

Reply:
I have kept all locking outside the library by design. I think this is best left up to the user to fix with a synchronization primitive of its choosing which might even be none with certain designs.

Reply:
For synchronisation inside an application I agree that we shouldn't impose any particular locking. But this also has to do with synchronisation *between* applications, that don't necessarily have any way of coordinating, other than libxdp. So I think solving this in a race-free way is in scope; it's basically the same issue as Maciej's patch[0] is addressing in the libbpf version of this code.

[0] https://lore.kernel.org/bpf/3a625c8b-6dc0-3933-25e5-f747197ae1f4@intel.com/T/#t